src/xdp-app-info-flatpak-private.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2024 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "xdp-app-info-private.h"
+
+struct _XdpAppInfoFlatpakClass
+{
+  XdpAppInfoClass parent_class;
+};
+
+#define XDP_TYPE_APP_INFO_FLATPAK (xdp_app_info_flatpak_get_type())
+G_DECLARE_FINAL_TYPE (XdpAppInfoFlatpak,
+                      xdp_app_info_flatpak,
+                      XDP, APP_INFO_FLATPAK,
+                      XdpAppInfo)
+
+gboolean xdp_app_info_flatpak_new (int          pid,
+                                   int          pidfd,
+                                   XdpAppInfo **out_app_info,
+                                   GError     **error);
```

Comment:
I think constructors like this should always return the object they're constructing:

```
XdpAppInfo *
xdp_app_info_flatpak_new (int      pid,
                          int      pidfd,
                          GError **error);
```

Reply:
That's not possible because this is basically a tri-state. While creating it, errors can occur, in which case no further attempt must be made to create another XdpAppInfo (because that might create a host XdpAppInfo for a flatpak app). It however can also be possible that the app just isn't e.g. a flatpak app and that snap and then host should be tried later.

The API is as good as it gets with those semantics because it follows the "returns FALSE when error is set" convention while also allowing to either create or not create an XdpAppInfo.

Reply:
Returning NULL meaning either "not a flatpak process" or "there was an error" is exactly how it worked before. In glib and a lot of other GNOME components there is a convention that the success/error state is returned and the error is just an additional piece of information. This change is so that it follows this convention.

I really don't understand the problem here. Out arguments are common. The error argument itself is such a thing.

Reply:
I think what I was trying to say is, a common pattern for this kind of situation is to define a custom GError, and use it. [Here's an example in Boatswain](https://gitlab.gnome.org/World/boatswain/-/blob/main/src/bs-device-manager.c?ref_type=heads#L67-77).

The `BS_STREAM_DECK_ERROR_UNRECOGNIZED` is defined specifically for the purpose of saying: this device is not a recognized Elgato device. Creating an object will fail (the purpose of GInitable is for failable constructors after all), and we use the extra bit of information - the error itself - to decide how to proceed.

Reply:
Sure, I'm now using a new `XDG_DESKTOP_PORTAL_ERROR_WRONG_APP_KIND` error to signal that the process isn't the of the app kind which was attempted to be created.